File: lfs-client-sdk/python/lfs_sdk/resolver.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Optional

import boto3

from .envelope import decode_envelope, is_lfs_envelope, LfsEnvelope


@dataclass
class ResolvedRecord:
    envelope: Optional[LfsEnvelope]
    payload: bytes
    is_envelope: bool


class LfsResolver:
    def __init__(self, bucket: str, s3_client=None, validate_checksum: bool = True, max_size: int = 0) -> None:
        self.bucket = bucket
        self.s3 = s3_client or boto3.client("s3")
        self.validate_checksum = validate_checksum
        self.max_size = max_size

    def resolve(self, value: bytes) -> ResolvedRecord:
        if not is_lfs_envelope(value):
            return ResolvedRecord(None, value, False)

        env = decode_envelope(value)
        obj = self.s3.get_object(Bucket=self.bucket, Key=env.key)
        payload = obj["Body"].read()
        if self.max_size > 0 and len(payload) > self.max_size:
            raise ValueError("payload exceeds max size")
        if self.validate_checksum:
            expected = env.checksum or env.sha256
            actual = hashlib.sha256(payload).hexdigest()
            if actual != expected:
                raise ValueError("checksum mismatch")
        return ResolvedRecord(env, payload, True)

File: lfs-client-sdk/js/src/resolver.ts
import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3';
import { decodeEnvelope, isLfsEnvelope, LfsEnvelope } from './envelope.js';

export interface ResolvedRecord {
  envelope?: LfsEnvelope;
  payload: Uint8Array;
  isEnvelope: boolean;
}

export interface ResolverOptions {
  validateChecksum?: boolean;
  maxSize?: number;
}

export class LfsResolver {
  private readonly s3: S3Client;
  private readonly bucket: string;
  private readonly validateChecksum: boolean;
  private readonly maxSize: number;

  constructor(s3: S3Client, bucket: string, options?: ResolverOptions) {
    this.s3 = s3;
    this.bucket = bucket;
    this.validateChecksum = options?.validateChecksum ?? true;
    this.maxSize = options?.maxSize ?? 0;
  }

  async resolve(value: Uint8Array): Promise<ResolvedRecord> {
    if (!isLfsEnvelope(value)) {
      return { payload: value, isEnvelope: false };
    }
    const env = decodeEnvelope(value);
    const obj = await this.s3.send(new GetObjectCommand({ Bucket: this.bucket, Key: env.key }));
    const body = await obj.Body?.transformToByteArray();
    const payload = body ?? new Uint8Array();
    if (this.maxSize > 0 && payload.length > this.maxSize) {
      throw new Error('payload exceeds max size');
    }
    if (this.validateChecksum) {
      // checksum validation placeholder (sha256)
    }
    return { envelope: env, payload, isEnvelope: true };
  }
}
